geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/iterators/CartesianProductIterable.scala

```diff
@@ -14,32 +14,51 @@ package org.locationtech.geomesa.utils.iterators
  * way to query the list size that is independent of the iterator itself.
  * (That is, asking for the size does not exhaust any iterator.)
  *
+ * NB:  The first sequence is the least significant; that is, it will
+ * increment fast while the last sequence is the most significant (will
+ * increment slowly).
+ *
  * @param seqs the list-of-lists whose items are to be recombined
  * @tparam T the type of items
  */
-case class CartesianProductIterable[T](seqs:Seq[Seq[T]]) extends Iterable[Seq[T]] {
-  // how many (total) combinations there will be in an iterator
-  lazy val expectedSize : Long = seqs.map(seq => seq.size.toLong).product
 
-  // create an iterator that will visit all possible combinations
-  // (every time you call this, you get a NEW iterator)
-  def iterator : Iterator[Seq[T]] = new Iterator[Seq[T]] {
-    // variable index that counts off the combinations
-    var index : Long = 0L
-    // current value (combination)
-    private def value : Seq[T] =
-      seqs.foldLeft(Seq[T](),index)((t,src) => t match { case (seqSoFar,idx) => {
-        val modulus : Int = src.size
-        val itemIndex : Int = (idx % modulus).toInt
-        val item : T = src(itemIndex)
-        (seqSoFar++Seq(item), idx / modulus)
-      }})._1
-    def next() : Seq[T] = {
-      val result = if (hasNext) value else null
-      index = index + 1L
+case class CartesianProductIterable(seqs: Seq[Seq[_]]) extends Iterable[Seq[_]] {
+  lazy val expectedSize: Int = seqs.map(_.size).product
```

Comment:
Should this still be long?

Reply:
Prolly; it was in the old version.  Good eye!

Reply:
I hope we don't every actually exceed Int.MaxValue here :P